### ISSUE TITLE
unionfs functionality 

### ISSUE THREAD

**Issue by ajkis** [billziss-gh/winfsp#86]
Is there a way to achieve unionfs functionality on windows.

**Comment by billziss-gh** [comment]
This is absolutely possible as unionfs is a relatively simple file system. However there is no such file system currently on Windows (that I am aware of).

If I was doing this I would probably start with the [passthrough](https://github.com/billziss-gh/winfsp/tree/master/tst/passthrough) (or [passthrough-fuse](https://github.com/billziss-gh/winfsp/tree/master/tst/passthrough-fuse)) file system. I would then add the unionfs functionality to it.

Here is a tutorial describing how to recreate the passthrough file system in case you are interested: https://github.com/billziss-gh/winfsp/wiki/WinFsp-Tutorial

Another option might be to port an existing unionfs file system to Cygwin/cygfuse. A search on github reveals various file systems that claim to provide unionfs functionality through FUSE.

**Comment by qnorsten** [comment]
Would love to see this happen also. Unfortunatly I lack the experience with filesystem program to make it happen, but might try to follow the passtrough tutorial just for fun and try and learn something. 

For unionfs-fuse systems on windows I found this, https://code.google.com/archive/p/winunionfs/ that uses Dokany, but have not tried it, so no idea how far they actually got/how functional it is. Perhaps it can be ported to winfsp instead and improved. Perhaps it might be easier to try and port the original unionfs-fuse project or start from scratch, in case any experienced developer decides to take up the challenge. At least winfsp seem to be an improvement over Dokany:) At least in my experience.

**Comment by fsgeek** [comment]
Union file systems are surprisingly difficult to get right (I was just discussing this with someone who spent considerable time building one in the past) because the name space rapidly becomes rather messy.

There are a couple of FUSE implementations floating around as well, though they are somewhat problematic.  Containers on Linux use various union style FS (or COW supporting FS) while native Windows containers (Server 2016) use file system filter drivers to achieve this functionality.  Those used by Docker appear to be kernel level VFS implementations for performance reasons.

**Comment by billziss-gh** [comment]
@qnorsten give it a try. There is a bit of a learning curve, but it is definitely worth learning about file systems.

@fsgeek I agree. Making a read-only unionfs is relatively easy, making one that is read-write is rather more difficult.

If you have any insider information on the gotchas when trying to build a unionfs, feel free to share them. Your expertise is invaluable.

----

Although I do not really have time for this at the moment, I contemplated last night doing something like it in Go with [cgofuse](https://github.com/billziss-gh/cgofuse). I thought I would allow the "lower" files to be placed in directories or tar/zip archives and I liked that Go has native support for those: [package archive](https://golang.org/pkg/archive/).

There is also @hasse69's [rar2fs](https://github.com/hasse69/rar2fs) which supports read-only access to rar archives and (I think) read-write access to directories. It was just ported to Windows: hasse69/rar2fs#75.

**Comment by billziss-gh** [comment]
@ajkis I am closing this as part of issue cleanup for the upcoming version `1.1` release.

I actually started building a "swiss knife" file system that includes unionfs functionality a few weekends ago (using Go :). Unfortunately I have very limited time currently and I am not able to devote any more time to it.

If I ever bring it to a point that is functional I will update this issue with a link. Closing for now as this issue is not WinFsp related.
